### ISSUE TITLE
array foreach implementation

### ISSUE THREAD

**Issue by medranocalvo** [flightaware/Tcl-bounties#10]
Hello,

here is an implementation of the `array foreach` improvement.  I changed the interface to be more similar to that of dict for: `array for {varName varName} arrayName { code }`, are you O.K. with that?  If so, I'll write required documentation changes and propose it for inclusion in Tcl core.  As a bonus, you can use this today, no need to wait on a new Tcl release.

Thank you,
Adrián

```tcl
proc ::tcl::array::for {variableNames arrayName code} {
    ::upvar 1 $arrayName arr;
    ::tailcall ::foreach $variableNames [::array get arr] $code
}
::namespace ensemble configure array -map \
    [dict replace [::namespace ensemble configure array -map] \
	 for ::tcl::array::for];

# Example usage:
array set myarr {
    x 1
    y 2
    z 3
}
array for {key val} myarr {
   puts "$key=$val";
}
```

**Comment by bovine** [comment]
A previously existing TIP was opened by karl years ago but never finished. The template might be reuable as the basis for this new TIP. http://www.tcl.tk/cgi-bin/tct/tip/421.html

Please note that the primary motivation of this enhancement is not just a syntactic improvement, but an actual performance improvement by eliminating the need to duplicate the array with [array get]

**Comment by mahmudsudo** [comment]
i would love to take on this

**Comment by sebres** [comment]
Regardless the complexity of this (NRE-capable and TEBC-compiled implementations of a loop)...

Take an attentive look at first sentence of the [README](https://github.com/flightaware/Tcl-bounties/blob/master/README.md):

> Due to the impact of the coronavirus, FlightAware's **Tcl Bounty program is temporarily on hold** until the business climate has stabilized.

I haven't heard that the statement has been retracted.
